File: ui-api/src/test/java/io/tipsters/oddsclient/WilliamHillClientTest.java
package io.tipsters.oddsclient;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.time.LocalDateTime;
import java.util.List;

import io.tipsters.testsupport.WilliamHillOddsFeedStub;
import io.tipsters.uiapi.client.williamhill.OddsClient;
import io.tipsters.uiapi.client.williamhill.UkOddsProvider;
import io.tipsters.uiapi.dto.CompetitionMatches;
import retrofit2.Retrofit;

import static com.google.common.collect.Sets.newHashSet;
import static org.assertj.core.api.Assertions.assertThat;

public class WilliamHillClientTest {
    private final int port = 19002;
    private final WilliamHillOddsFeedStub oddsApiStub = new WilliamHillOddsFeedStub(port);
    private final Retrofit httpClientBuilder = new Retrofit.Builder()
            .baseUrl("http://localhost:" + port)
            .build();

    @Before
    public void setUp() {
        oddsApiStub.start();
    }

    @After
    public void tearDown() {
        oddsApiStub.stop();
    }

    @Test
    public void returnsTheOddsFeed() throws IOException {
        OddsClient.Uk underTest = httpClientBuilder.create(OddsClient.Uk.class);

        oddsApiStub.willReturnTheOddsXMLFeed();

        UkOddsProvider oddsProvider = new UkOddsProvider(underTest);
        List<CompetitionMatches> competitions = oddsProvider.odds(newHashSet("Scottish Premiership"), LocalDateTime.MIN, LocalDateTime.MAX);

        assertThat(competitions.get(0).getCompetition()).isEqualTo("Scottish Premiership");
        assertThat(competitions.get(0).getMatches().get(0).getHome()).isEqualTo("Inverness");
        assertThat(competitions.get(0).getMatches().get(0).getAway()).isEqualTo("Hearts");
    }
}

File: odds-client/src/test/kotlin/io/tipsters/oddsclient/WilliamHillClientTest.kt
package io.tipsters.oddsclient

import io.tipsters.common.data.MatchesByCompetition
import io.tipsters.oddsfeedclient.WilliamHillUkOdds
import io.tipsters.oddsfeedclient.oddsprovider.WilliamHillUkOddsProvider
import io.tipsters.testsupport.stub.WilliamHillOddsFeedStub
import org.junit.After
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import retrofit2.Retrofit
import java.time.LocalDateTime
import org.hamcrest.CoreMatchers.`is` as eq

class WilliamHillClientTest {
    val port = 19002
    val oddsApiStub = WilliamHillOddsFeedStub(port)
    val httpClientBuilder = Retrofit.Builder()
            .baseUrl("http://localhost:$port/")
            .build();

    val underTest = httpClientBuilder.create(WilliamHillUkOdds::class.java)

    @Before
    fun setUp() {
        oddsApiStub.start()
    }

    @After
    fun tearDown() {
        oddsApiStub.stop()
    }

    @Test
    fun returnsTheOddsFeed() {
        oddsApiStub.willReturnTheOddsXMLFeed()

        val oddsProvider = WilliamHillUkOddsProvider(underTest)
        val competitions: List<MatchesByCompetition> = oddsProvider.odds(setOf("Scottish Championship"), LocalDateTime.MIN, LocalDateTime.MAX)

        assertThat(competitions[0].competition, eq("Scottish Championship"))
        assertThat(competitions[0].matches[0].home, eq("Hibernian"))
        assertThat(competitions[0].matches[0].away, eq("Falkirk"))
    }
}
